docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,349 @@
+```
+
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.unknown.artifact.v1",
+  "content": {
+    "config": {
+      "mediaType": "application/vnd.oci.empty.v1+json",
+      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
+      "size": 2,
+      "data": "e30="
+    },
+    "layers": [
+      {
+        "mediaType": "application/vnd.oci.image.layer.v1.tar",
+        "digest": "sha256:6cb759c4296e67e35b0367f3c0f51dfdb776a0c99a45f39d0476e43d82696d65",
+        "size": 14477,
+        "annotations": {
+          "org.opencontainers.image.title": "sbom.spdx"
+        }
+      },
+      {
+        "mediaType": "application/vnd.oci.image.layer.v1.tar",
+        "digest": "sha256:54c0e84503c8790e03afe34bfc05a5ce45c933430cfd9c5f8a99d2c89f1f1b69",
+        "size": 6639,
+        "annotations": {
+          "org.opencontainers.image.title": "scan-test-verify-image.json"
+        }
+      }
+    ],
+  "annotations": {
+    "org.opencontainers.image.created": "2023-12-15T09:41:54Z"
+  }
+},
+}
+```
+
+> [!NOTE]
+> - `--output -` and `--format` can not be used at the same time due to conflicts.
+
+## Scenarios
+
+### Scripting
+Alice is a developer who wants to batch operations with ORAS in her Shell script. In order to automate a portion of her workflow, she would like to obtain the image digest from the JSON output objects produced by the `oras push` command and then use shell variables or utilities like [xargs](https://en.wikipedia.org/wiki/Xargs) to enable an ORAS command to act on the output of another command and perform further steps. In this way, she can chain commands together. For example, she can use `oras attach` to attach an SBOM to the image using its image digest as an argument outputted from `oras push`.
+Alice is a developer who wants to batch operations with ORAS in her Shell script. In order to automate a portion of her workflow, she would like to obtain the image digest from the JSON output objects produced by the `oras push` command and then use shell variables or utilities like [xargs](https://en.wikipedia.org/wiki/Xargs) to enable an ORAS command to act on the output of another command and perform further steps. In this way, she can chain commands together. For example, she can use `oras attach` to attach an SBOM to the image using its image digest as a argument outputted from `oras push`.
+
+For example, push an artifact to a registry and generate the artifact reference in the standard output. Then, attach an SBOM to the artifact using the artifact reference (`$REGISTRY/$REPO@$DIGEST`) outputted from the first command. Finally, sign the attached SBOM with another tool against the reference of the SBOM file (`$REGISTRY/$REPO@$DIGEST`) that was obtained in the proceeding step.
+
+- Use shell variables on Unix
+
+```bash
+REFERENCE_A=$(oras push $REGISTRY/$REPO:$TAG hello.txt --format '{{.reference}}')
```

Comment:
Looking at tools like Trivy and Grype, it seems that they separate the `--format` from the template. Both those tools have `--format` switch that takes values like `json`, `spdx`, etc and one of those values is `template`. If `template` is selected as a value, then additional switch `--template` is required to specify the Go template.

Reply:
Also, looking at skopeo, the `-f` switch is only for Go template (which can result in JSON) but it will be very complex template to create table, CSV etc.

Reply:
Actually having a `-o template --template ...` does align even with kubectl 

# Return only the phase value of the specified pod
  kubectl get -o template pod/web-pod-13je7 --template={{.status.phase}}

Reply:
@toddysm 
It would be much reasonable to align the design principles of formatted output with container management CLI tools like `kubectl`, `docker`. Trivy and Grape are vulnerability scanning tools that may have different design.

`kubectl` also puts the "Go template" along with different output formats under its `--output` flag. ORAS borrowed its idea and simplied the usage. See https://kubernetes.io/docs/reference/kubectl/quick-reference/#formatting-output

`oras` has almost the same behavior of formatted output control as `docker`. See https://docs.docker.com/reference/cli/docker/image/ls/#format.

If we separate the data format and Go template into two different flags, i.e. `--format <data_format>` and `--template <go_template>`, it would be much more complicated because ORAS requires to specify two flags in some cases.

Reply:
@FeynmanZhou as per @sajayantony comment above, the two flags are also used in `kubectl` Having two flags is not a complication but better usability IMHO. Overloading flags adds to confusion, and it is considered not a good experience practice.

Reply:
I think the main point here is that don't overload the `--format` and have a separate argument for `--template`. Basically avoid over load bearing arguments.

Reply:
OK, I went ahead and did the analysis on the common tools mentioned above: `skopeo`, `kubectl`, `docker`, `trivy`, and `grype` and here is the summary.

# `skopeo`
First of all, `skopeo` has no `--output` switch at all. Which means that you cannot control how the output is formated (JSON, YAML, table, tree, TXT, CSV, etc.) and the supported formats are either JSON or TXT based on the comand. Users are left to the decision of the `skopeo` implementers how to format the output. Commands that have meaningful output and can be used for more structured parsing are `inspect`, `list-tags`, and `manifest-digest`. The first two return JSON from, while the latter returns the digest as single value.

Looking at the `-f, --format` switch, it is available for the following commands:
- `copy`
   In this case it specifies what should be the target manifest format and has nothing to do with the format of the output. It expects values like `oci`, `v2s1, and `v2s2` (nothing to do with output format as we define it).
- `inspect`
  This one uses ONLY Go template format, and IMHO is more intended to filter data than to format the output as it seems from the example.
- `sync`
   In this case it specifies what should be the target manifest format and has nothing to do with the format of the output. It expects values like `oci`, `v2s1, and `v2s2` (nothing to do with output format as we define it). Same as `copy`.

As an overall opinion, I find the `skopeo` switches confusing and not very intuitive as the example with `-f` shows above, and I don't think their design should be used as an user-friendly example.

# `kubectl`
`kubectl` uses `-o, --output` that is intended to spefify how the output should be formatted. This switch expects various options like `json`, `yaml`, etc including `template`, and `go-template` (not sure what is the difference between those two BTW). It also uses the `--template` switch if the `-o` switch is set to `template` or (I believe) `go-template` (or their respective file path options).

Using the above two switches, one has very flexible and user-friendly approach to control the output. The `-f` switch is with a very clear purpose - either standard, well-known formats like `json`, or custom controlled that require the `--template` switch.

Note that `kubectl` is not intended to safe (local) files as a result of the commands. You can still pipe the output (as specified and formatted with the two switches above) and save the output to a file (or use in another command). It does have `-f` switch that is intended as an input for identifying the resource. It also has `--field-selector`, which is used to filter the resources on which the command will operate (not exactly the filtering we want to have in ORAS but I have a point below).

While I cannot say that all the `-o` options are clear to me (not sure what `name` is, for example), there is clear separation between switches and their purpose and they do not overload the `-o` with enumerated formats and free form templates.

# `docker`
`docker` is the one that overloads the `--format` and has the options either using `json` or Go template. It does not have the flexibility to load templates from files or any other options like `yaml`, `table`, `tree` or what have you. In general - much less flexible than `kubectl`.

# `trivy`
`trivy` is aligned with what `kubectl` has. They use `-f, --format` for the enumeration of possible formats as well as `--template` in the case of a Go template selected in `-f`. They also use `-o, --output` to save the result to a file (something that `kubectl` doesn't have). Note, though, that Trivy's `-o` is to save the output of the command and not save the downloaded content from the server as it may be the case with ORAS. 

# `grype`
`grype` doesn't have `--format` because they use the short-hand switch `-f` for other purposes. Instead, they use the `-o, --output` for that purpose but as with `trivy` and `kubectl` they require `--template` switch if `template` is specified in `-o` (only path to a Go template file is accepted). Unlike `trivy` they don't have a switch to save the output.

# Summary
IMHO, the most user-friendly and intuitive approaches are by `kubectl` and `trivy`. In both of those the output format choices and the Go template are specified using different switches. I find `trivy`'s enumerated list more clear than `kubectl` one but this is beyond the point.

Nevertheless, having clear purpose of the switches makes the CLI much clearer and flexible than overloading switches (as we see with `skopeo`)

IMHO, I think that we need to have the following switches and use those consistently across all the commands in ORAS:
- `-f, --format` for the enumeration of possible command output formats (`json`, `yaml`, `tree`, `template`, etc.)
- We currently use `-o, --output` for the destination of the downloaded content from the server. This can be the same as the command output (in case of getting the manifest) or different (in the case of saving the blob). Maybe a better switch would be `-d, --destination`to avoid confusion with the command output.
- `--template` for specifying the Go template only if `-f` is set to `template` (or eventually `template-file`, that is not available right now)

Reply:
I am okay with separating the flags in ORAS, but I need to point out that `kubectl` does overload -o/--output with output format (json/yaml/table/name) and go template.

Reply:
@qweeah I see what you mean but I find this confusing. Isn't this the same as `-o template` and the `--template {{.spec}}`?

Reply:
@toddysm indeed. `kubectl` makes `--template xxx` equivalent to `-o template=xxx`, which we need to avoid since we do not want to overloaded flags.

Reply:
Thanks @toddysm for the detailed summary. 

After tested `kubectl` and `trivy`, I found these two problems that we may need to notice:

 1. `kubectl` provides three UX to allow use Go template and **the second UX overloads the go template argument with other enumerated output formats**. The second one is almost the same as the current design of oras `--format <go_template>`:

```
kubectl get svc/kubernetes -o template --template='{{.metadata}}'
```
is equivalent to
```
kubectl get svc/kubernetes -o template='{{.metadata}}'
```
is equivalent to
```
kubectl get svc/kubernetes  --template='{{.metadata}}'
```

2. Neither `kubectl` nor `trivy` supports specifying a data format like `json/yaml/table` and manipulate the data with `--template <go_template_function>` at the same time. The printed output of using latter flag `--template` will be overwritten by former one. 

For example, [trivy](https://aquasecurity.github.io/trivy/v0.50/docs/configuration/reporting/#custom-template) allows to manipulate data with `--template <go_template_function>` only when specifying `--format template`:

```
trivy image --format template --template <go_template_function> golang:1.19-alpine
```

### Summary

If we want to avoid overloading and learn good aspects from `kubectl` and `trivy`, it should be reasonable to split it into two flags `--format` and `--template`. The UX will be like the following example:

```
oras push $REGISTRY/$REPO:$TAG hello.txt --format <go-template/json> --template <go_template/file_path>
```

Reply:
Template file could be another arg but maybe later.

Reply:
Related to https://github.com/oras-project/oras/pull/1199#discussion_r1578838180, @qweeah and @FeynmanZhou  the difference between `kubectl` and our proposal is that `kubectl` requires the word `template` (and `=`) before the template is specified in the `-o` flag. This is much clearer than just pasting the template in the `-o` flag. While I can see how this may be convenient for short templates, I still think it is inconsistent from experience point of view. 
Regarding 2. from @FeynmanZhou above - I agree that this is the case. The filtering is a separate problem that I think we should tackle in a new issue.

Reply:
Had a discussion with @toddysm . The decision is to have two separate flags `--format <data_format>` and `--go-template <go_template_function>` because it will not overload two operations in the same flag. 
I will update the Spec to reflect the changes as we discussed above.

Reply:
Please align the flag values with kubectl. 
I think it’s —format go-template.

Reply:
`--format go-template` makes more sense to me and is a better generic design in general.

Reply:
I have separated the `--format` flag into two different flags `--format` and `--template` at [line 28 and 29](https://github.com/oras-project/oras/pull/1199/files#diff-ddd85908fb8fee8c57a3726cb042b0572e9b4b333000356816268399ced78cccR28) and updated all samples in this Spec: 

> - Use `--format json|tree|table|go-template='{{ GO_TEMPLATE }}'` to format metadata output of ORAS commands into different formats including prettified JSON, tree, table view, and Go template. It supports computing figures within the template using [Sprig](http://masterminds.github.io/sprig/) functions. Make `-f` as an alias of `--format`.
> - Use `--template '{{ GO_TEMPLATE }}'` to compute and manipulate the output data using Go template based on the chosen data format. To avoid ambiguous, this flag can only be used along with `--format <DATA_FORMAT>`.